src/main/java/com/ning/maven/plugins/dependencyversionscheck/AbstractDependencyVersionsMojo.java

```diff
@@ -487,23 +544,25 @@ private boolean isVisible(final String scope, final String [] validScopes)
      */
     private void addToResolutionMap(final Map resolutionMap, final VersionResolution resolution)
     {
-        List resolutions = (List) resolutionMap.get(resolution.getDependencyName());
-        if (resolutions == null) {
-            resolutions = new ArrayList();
-            resolutionMap.put(resolution.getDependencyName(), resolutions);
-        }
-
-        for (Iterator it = resolutions.iterator(); it.hasNext(); ) {
-            final VersionResolution existingResolution = (VersionResolution) it.next();
-            // TODO: It might be reasonable to fail the build in this case. However, I have yet to see
-            // this message... :-) 
-            if (!existingResolution.getActualVersion().equals(resolution.getActualVersion())) {
-                LOG.warn("Dependency '{} expects version '{}' but '{}' already resolved to '{}'!", 
-                         new Object [] { resolution.getDependencyName(), resolution.getActualVersion(), existingResolution.getDependencyName(), existingResolution.getActualVersion() });
+        synchronized (resolution.getDependencyName().intern()) {
```

Comment:
This kind of locking is bad practice. The string object is global, so for example, if multiple instances of this plugin is running (which is the case for a parallel build), they will block each other. You could also have something unrelated in the same JVM messing with the same lock.

Guava `Cache` can be used here to replace the map: http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/cache/Cache.html#get(K,%20java.util.concurrent.Callable)

It has a per-key lock around the loading method.

https://github.com/google/guava/wiki/CachesExplained